### ISSUE TITLE
Review the indexation of properties and make it more flexible in terms of configuration

### ISSUE THREAD

**Issue by beligum** [republic-of-reinvention/com.stralo.framework#50]

**Comment by beligum** [comment]
Decided to drop support for Lucene for full-text indexing and use Solr embedded by default, because it offers us a lot more flexibility (and made some decisions for us)

**Comment by beligum** [comment]
Unifying resource summary interfaces into one:
* IndexEntry
* SummarizedResource
* ResourceInfo
* AutocompleteSuggestion

**Comment by beligum** [comment]
Thinking a bit deeper on this. Note that our definition of a "Resource" is:
`Something our system can serve to the end user by means of a public URI`
So, by extension, we're looking for an API to serve a subset of a Resource, or a Proxy.
So let's introduce a new concept: a `ResourceProxy`

**Comment by beligum** [comment]
After the previous change, the indexing of RDF values (remote or local) is now implemented as follows:
-  if it's a literal value, put the value in the json object
- If the value is a resource, we'll store its reference:
  - if its a sub-resource, the data of it will be in the RDF model and we need to 'instantiate' these in the JSON tree by saving the reference to the right property now and create + attach the sub-object later on (see below)
  - if it's a reference to a (local or external) resource that has an endpoint, we'll store some metadata about that resource in this JSON object (eg. for sorting/filtering on the human readable label of the resource instead of its URI)
  - if it's a reference to a resource without an endpoint,

**Comment by beligum** [comment]
When considering alternatives, this is a possibility that comes to mind:
```json
{
  "image" : null,
  "language" : "en",
  "resource" : "/resource/1106177060466942658",
  "parentUri" : null,
  "uri" : "/en/blah",
  "label" : "rdfs label test 9",
  "description" : "A title\n        Here you can type some text",
  "typeOf" : "ror:Page",
  "rdf:type" : [ "http://www.reinvention.be/ontology/Page", {
    "uri" : "http://www.reinvention.be/ontology/Page"
  } ],
  "rdfs:label" : "rdfs label test 9",
  "meta:created" : "2019-04-04T09:08:35.736Z",
  "meta:creator" : [ "/users/2", {
    "uri" : "/users/2"
  } ],
  "meta:modified" : "2019-04-13T15:00:08.816Z",
  "meta:contributor" : [ "/users/2", {
    "uri" : "/users/2"
  } ],
  "ror:text" : "A title Here you can type some text"
}
```

**Comment by beligum** [comment]
Implemented the above thought in the final indexer, both for sub-objects and references to other (local or external) resources.

**Comment by beligum** [comment]
Reopening this issue because the above decision wasn't all too great. The problem is that we should allow the index to be different from the triple store. On top, Solr advises to denormalize as much as possible and we also need a cleaner datastructure to send back in our upcoming public API.

So we decided to implement this approach:

If the value is a resource, we'll store its reference:
* if its a sub-resource, index it as a subnode of the parent under it's normal property name
* if it's a reference to a (local or external) resource that has an endpoint, store the proxy object as a subnode of the parent under it's normal property name (so no more _proxy fields)
* if it's a reference to a resource without an endpoint, create a subnode with and identifier (either "uri", "resource" or both, tbd)

**Comment by beligum** [comment]
Seems to work okay now, closing until further notice...

**Comment by beligum** [comment]
Note that we needed to introduce a new internal field "_id", because when creating a nested proxy object inside a document, the uri of that proxy document would be the same as the uri of the original document, resulting in two Solr docs with the same uri, eg. the same uniqueKey (if we configure uri as the unique key).
This introduces all kinds of problems and is just wrong. So we created a truly unique autogenerated UUID field, that's marked as the uniqueKey in SolrConfig
